firedrake/slate/slac/kernel_builder.py

```diff
@@ -379,3 +391,418 @@ def needs_mesh_layers(self):
+                    for i, shape in enumerate(shapes):
+                        # Preserves relation between slate Tensor and gem Indexed tensor
+                        # Subblocks of mixed Tensors have distint space, offset index and shape,
+                        # but share the gem Node for the tensor
+                        cinfo = CoefficientInfo(space_index=i,
+                                                offset_index=offset,
+                                                shape=shape,
+                                                vector=tensor,
+                                                local_temp=local_temp,
+                                                function=f)
+                        coeff_vecs.setdefault(((shape, offset), tensor.is_mixed), []).append(cinfo)
+                        offset += shape
+
+                    seen_coeff.add(function)
+
+        # Collect all coefficients which are not wrapped by an AssembledVector (e.g. of Tensor(L))
+        for i, c in enumerate(expression.coefficients()):
+            if c not in seen_coeff and (type(c) == Constant or type(c) == Function or type(c) == Coefficient):
+                element = c.ufl_element()
+                if type(element) == MixedElement:
+                    for j, c_ in enumerate(c.split()):
+                        name = "w_{}_{}".format(i, j)
+                        extra_coefficients.extend([(c, name)])
+                        self.args_extents.setdefault(name, index_extent(c_))
+                else:
+                    name = "w_{}".format(i)
+                    extra_coefficients.extend([(c, name)])
+                    self.args_extents.setdefault(name, index_extent(c))
+                seen_coeff.add(c)
+
+        self.expression = expression
+        self.tsfc_parameters = tsfc_parameters
+        self.temps = temps
+        self.gem_loopy_dict = gem_loopy_dict
+        self.ref_counter = counter
+        self.expression_dag = expression_dag
+        self.coefficient_vecs = coeff_vecs
+        self.extra_coefficients = extra_coefficients
+        self._setup()
+
+    @cached_property
+    def context_kernels(self):
+        r"""Gathers all :class:`~.ContextKernel`\s containing all TSFC kernels,
+        and integral type information.
+        """
+
+        cxt_list = [compile_terminal_form(expr, prefix="subkernel%d_" % i,
+                                          tsfc_parameters=self.tsfc_parameters, coffee=False)
+                    for i, expr in enumerate(self.temps)]
+
+        cxt_kernels = [cxt_k for cxt_tuple in cxt_list
+                       for cxt_k in cxt_tuple]
+        return cxt_kernels
+
+    def shape(self, tensor):
+        """ A helper method to retrieve tensor shape information.
+        In particular needed for the right shape of scalar tensors.
+        """
+        if tensor.shape == ():
+            return (1, )  # scalar tensor
+        else:
+            return tensor.shape
+
+    def _setup(self):
```

Comment:
I'm not really a fan of these mega-functions. It's very hard to see what is going on, and they're hard to extend.